File: backend/app/routers/gardens.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.database import get_db
from app.dependencies import get_current_user
from app.models.garden import Garden
from app.models.plant import Plant
from app.models.user import User
from app.models.zone import Zone
from app.schemas.garden import (
    GardenCreate,
    GardenDetailResponse,
    GardenResponse,
    GardenUpdate,
    ZoneCreate,
    ZoneResponse,
    ZoneUpdate,
)

router = APIRouter(prefix="/api/v1/gardens", tags=["gardens"])


@router.post("", response_model=GardenResponse, status_code=status.HTTP_201_CREATED)
async def create_garden(
    data: GardenCreate,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    garden = Garden(
        user_id=user.id,
        name=data.name,
        description=data.description,
        location_lat=data.location_lat,
        location_lng=data.location_lng,
    )
    db.add(garden)
    await db.commit()
    await db.refresh(garden)
    return garden


@router.get("", response_model=list[GardenResponse])
async def list_gardens(
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(Garden).where(Garden.user_id == user.id).order_by(Garden.created_at)
    )
    return result.scalars().all()


@router.get("/{garden_id}", response_model=GardenDetailResponse)
async def get_garden(
    garden_id: uuid.UUID,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(Garden)
        .options(selectinload(Garden.zones))
        .where(Garden.id == garden_id, Garden.user_id == user.id)
    )
    garden = result.scalar_one_or_none()
    if not garden:
        raise HTTPException(status_code=404, detail="Garden not found")

    count_result = await db.execute(
        select(func.count()).select_from(Plant).where(Plant.garden_id == garden_id, Plant.is_active == True)
    )
    plant_count = count_result.scalar() or 0

    return GardenDetailResponse(
        **{c.key: getattr(garden, c.key) for c in Garden.__table__.columns},
        zones=[ZoneResponse.model_validate(z) for z in garden.zones],
        plant_count=plant_count,
    )


@router.patch("/{garden_id}", response_model=GardenResponse)
async def update_garden(
    garden_id: uuid.UUID,
    data: GardenUpdate,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(Garden).where(Garden.id == garden_id, Garden.user_id == user.id)
    )
    garden = result.scalar_one_or_none()
    if not garden:
        raise HTTPException(status_code=404, detail="Garden not found")

    for key, value in data.model_dump(exclude_unset=True).items():
        setattr(garden, key, value)
    await db.commit()
    await db.refresh(garden)
    return garden


@router.delete("/{garden_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_garden(
    garden_id: uuid.UUID,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(Garden).where(Garden.id == garden_id, Garden.user_id == user.id)
    )
    garden = result.scalar_one_or_none()
    if not garden:
        raise HTTPException(status_code=404, detail="Garden not found")
    await db.delete(garden)
    await db.commit()


# ---- Zones ----

@router.post("/{garden_id}/zones", response_model=ZoneResponse, status_code=status.HTTP_201_CREATED)
async def create_zone(
    garden_id: uuid.UUID,
    data: ZoneCreate,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(Garden).where(Garden.id == garden_id, Garden.user_id == user.id)
    )
    if not result.scalar_one_or_none():
        raise HTTPException(status_code=404, detail="Garden not found")

    zone = Zone(garden_id=garden_id, **data.model_dump())
    db.add(zone)
    await db.commit()
    await db.refresh(zone)
    return zone


@router.get("/{garden_id}/zones", response_model=list[ZoneResponse])
async def list_zones(
    garden_id: uuid.UUID,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(Garden).where(Garden.id == garden_id, Garden.user_id == user.id)
    )
    if not result.scalar_one_or_none():
        raise HTTPException(status_code=404, detail="Garden not found")

    zones = await db.execute(
        select(Zone).where(Zone.garden_id == garden_id).order_by(Zone.sort_order)
    )
    return zones.scalars().all()


@router.patch("/{garden_id}/zones/{zone_id}", response_model=ZoneResponse)
async def update_zone(
    garden_id: uuid.UUID,
    zone_id: uuid.UUID,
    data: ZoneUpdate,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(Garden).where(Garden.id == garden_id, Garden.user_id == user.id)
    )
    if not result.scalar_one_or_none():
        raise HTTPException(status_code=404, detail="Garden not found")

    zone_result = await db.execute(
        select(Zone).where(Zone.id == zone_id, Zone.garden_id == garden_id)
    )
    zone = zone_result.scalar_one_or_none()
    if not zone:
        raise HTTPException(status_code=404, detail="Zone not found")

    for key, value in data.model_dump(exclude_unset=True).items():
        setattr(zone, key, value)
    await db.commit()
    await db.refresh(zone)
    return zone


@router.delete("/{garden_id}/zones/{zone_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_zone(
    garden_id: uuid.UUID,
    zone_id: uuid.UUID,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(Garden).where(Garden.id == garden_id, Garden.user_id == user.id)
    )
    if not result.scalar_one_or_none():
        raise HTTPException(status_code=404, detail="Garden not found")

    zone_result = await db.execute(
        select(Zone).where(Zone.id == zone_id, Zone.garden_id == garden_id)
    )
    zone = zone_result.scalar_one_or_none()
    if not zone:
        raise HTTPException(status_code=404, detail="Zone not found")
    await db.delete(zone)
    await db.commit()

File: mobile/src/services/gardens.ts
import { api } from './api';

export interface Garden {
  id: string;
  user_id: string;
  name: string;
  description: string | null;
  location_lat: number | null;
  location_lng: number | null;
  hardiness_zone: string | null;
  soil_type: string | null;
  privacy: string;
  zones: Zone[];
  plant_count: number;
}

export interface Zone {
  id: string;
  garden_id: string;
  name: string;
  light_type: string | null;
  zone_type: string | null;
  sort_order: number;
  position_x: number;
  position_y: number;
  width: number;
  height: number;
}

export interface Plant {
  id: string;
  garden_id: string;
  zone_id: string | null;
  species_id: string | null;
  custom_name: string;
  planting_date: string | null;
  location_detail: string | null;
  container_type: string | null;
  notes: string | null;
  health_status: 'healthy' | 'warning' | 'critical' | 'unknown';
  is_active: boolean;
  position_x: number;
  position_y: number;
  cover_photo_url: string | null;
}

// Gardens
export async function createGarden(data: { name: string; description?: string; location_lat?: number; location_lng?: number }) {
  const { data: garden } = await api.post<Garden>('/api/v1/gardens', data);
  return garden;
}

export async function listGardens() {
  const { data } = await api.get<Garden[]>('/api/v1/gardens');
  return data;
}

export async function getGarden(gardenId: string) {
  const { data } = await api.get<Garden>(`/api/v1/gardens/${gardenId}`);
  return data;
}

export async function updateGarden(gardenId: string, updates: Partial<Garden>) {
  const { data } = await api.patch<Garden>(`/api/v1/gardens/${gardenId}`, updates);
  return data;
}

// Zones
export async function createZone(gardenId: string, data: { name: string; zone_type?: string; light_type?: string; position_x?: number; position_y?: number; width?: number; height?: number }) {
  const { data: zone } = await api.post<Zone>(`/api/v1/gardens/${gardenId}/zones`, data);
  return zone;
}

export async function listZones(gardenId: string) {
  const { data } = await api.get<Zone[]>(`/api/v1/gardens/${gardenId}/zones`);
  return data;
}

// Plants
export async function createPlant(gardenId: string, data: { custom_name: string; species_id?: string; zone_id?: string; position_x?: number; position_y?: number }) {
  const { data: plant } = await api.post<Plant>(`/api/v1/gardens/${gardenId}/plants`, data);
  return plant;
}

export async function listPlants(gardenId: string, filters?: { zone_id?: string; health_status?: string }) {
  const { data } = await api.get<Plant[]>(`/api/v1/gardens/${gardenId}/plants`, { params: filters });
  return data;
}

export async function getPlant(plantId: string) {
  const { data } = await api.get<Plant>(`/api/v1/plants/${plantId}`);
  return data;
}

export async function updatePlant(plantId: string, updates: Partial<Plant>) {
  const { data } = await api.patch<Plant>(`/api/v1/plants/${plantId}`, updates);
  return data;
}
